README.md
# Kakapo ![partyparrot](http://cultofthepartyparrot.com/sirocco.gif)
[![Language: Swift](https://img.shields.io/badge/lang-Swift-yellow.svg?style=flat)](https://developer.apple.com/swift/)
[![Build Status](https://travis-ci.org/devlucky/Kakapo.svg?branch=master)](https://travis-ci.org/devlucky/Kakapo)
[![Version](https://img.shields.io/cocoapods/v/Kakapo.svg?style=flat)](http://cocoapods.org/pods/Kakapo)
[![DocCov](https://img.shields.io/cocoapods/metrics/doc-percent/Kakapo.svg)](http://cocoadocs.org/docsets/Kakapo)
[![codecov](https://codecov.io/gh/devlucky/Kakapo/branch/master/graph/badge.svg)](https://codecov.io/gh/devlucky/Kakapo)
[![codebeat badge](https://codebeat.co/badges/69a42ece-740c-4a29-b25a-598deaf61fca)](https://codebeat.co/projects/github-com-devlucky-kakapo)
[![License](https://img.shields.io/cocoapods/l/Kakapo.svg?style=flat)](http://cocoapods.org/pods/Kakapo)
[![Platform](https://img.shields.io/cocoapods/p/Kakapo.svg?style=flat)](http://cocoapods.org/pods/Kakapo)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)

> Dynamically Mock server behaviors and responses.

## Contents
- [Why Kakapo?](#why-kakapo)
- [Features](#features)
- [Installation](#installation)
- [Usage](#usage)
  - [Serializable protocol](#serializable-protocol)
  - [Router: Register and Intercept](#router---register-and-intercept)  
    - [Third-party network libraries](#third-party-libraries)
  - [Leverage the Store - Dynamic mocking](#leverage-the-store---dynamic-mocking)
  - [CustomSerializable](#customserializable)
  - [JSONAPI](#jsonapi)
  - [Expanding Null values with Property Policy](#expanding-null-values-with-property-policy)
  - [Key customization - Serialization Transformer](#key-customization---serialization-transformer)
  - [Customize responses with ResponseFieldsProvider](#customize-responses-with-responsefieldsprovider)
- [Roadmap](#roadmap)
- [Examples](#examples)
- [Authors](#authors)

Kakapo is a dynamic mocking library. It allows you to replicate your backend APIs and logic.  
With Kakapo you can easily prototype your application based on your API specifications.

## Why Kakapo?

A common approach when testing network requests is to stub them with fake network responses from local files or recorded requests. This has some disadvantages:

- All files need to be updated when the APIs are updated.
- Lots of files have to be generated and included in the project.
- Are just static responses that can only be used for unit tests since they don't reflect backend behaviors and state.

While still this approach may work good, Kakapo will be a game changer in your network tests: it will give you complete control when it comes to simulating backend behaviors. Moreover, is not just unit testing: you can even take a step further and prototype your application before having a real service behind!  
With Kakapo you can just create Swift structs/classes/enums that are automatically serialized to JSON.


> 7 billion people on Earth
>
> Less than 150 Kakapo
>
> Time is critical [donate to Kakapo recovery](http://kakaporecovery.org.nz/adopt-a-kakapo/)

## Features

  * Dynamic mocking
  * Prototyping
  * Swift 3.0 compatible (from version 2.0.0, master branch)
  * Swift 2.2 and 2.3 compatible (from version 1.0.0, branch feature/legacy-swift)
  * Compatible with [![Platform](https://img.shields.io/cocoapods/p/Kakapo.svg?style=flat)](http://cocoapods.org/pods/Kakapo)
  * Protocol oriented and pluggable
  * Fully customizable by defining custom serialization and custom responses
  * Out-of-the-box serialization
  * [JSONAPI](http://jsonapi.org) support

## Installation

Using [CocoaPods](http://cocoapods.org/):

```ruby
use_frameworks!
pod 'Kakapo'
```

You can use the "Serialization" subspec to only use the object->JSON serialization features of Kakapo.  

Using [Carthage](https://github.com/Carthage/Carthage):

```
github "devlucky/Kakapo"
```

## Usage

> NOTE: The project also contains a [README.playground](https://github.com/devlucky/Kakapo/tree/master/README.playground).
> Check it out to see some examples of the key features of Kakapo.

Kakapo is made with an easy-to-use design in mind. To quickly get started, you can create a `Router` that intercepts network requests like this:

```Swift
let router = Router.register("http://www.test.com")
router.get("/users") { request in
  return ["id" : 2, "name": "Kakapo"]
}
```

You might be wondering where the dynamic part is; here is when the different modules of Kakapo take place:

```Swift
let store = Store()
store.create(User.self, number: 20)

router.get("/users") { request in
  return store.findAll(User.self)
}
```

Now, we've created 20 random `User` objects and mocked our request to return them.  

Let's get a closer look to the different features:

### Serializable protocol

Kakapo uses the `Serializable` protocol in order to serialize objects to JSON. *Any type* can be serialized as long as it conforms to this protocol:

```Swift
struct User: Serializable {
  let name: String
}

let user = User(name: "Alex")
let serializedUser = user.serialized()
//  -> ["name": "Alex"]
```

Also, standard library types are supported: this means that `Array`, `Dictionary` or `Optional` can be serialized:

```Swift
let serializedUserArray = [user].serialized()
// -> [["name": "Alex"]]
let serializedUserDictionary = ["test": user].serialized()
// -> ["test": ["name": "Alex"]]
```

### Router - Register and Intercept

Kakapo uses `Router`s in order to keep track of the registered endpoints that have to be intercepted.  
You can match *any* relative path from the registered base URL, as long as the components are matching the request's components. You can use wildcard components:

```Swift
let router = Router.register("http://www.test.com")

// Will match http://www.test.com/users/28
router.get("/users/:id") { ... }

// Will match http://www.test.com/users/28/comments/123
router.get("/users/:id/comments/:comment_id") { ... }
```

The handler will have to return a `Serializable` object that will define the response once the URL of a request is matched.
When a `Router` intercepts a request, it automatically serializes the `Serializable` object returned by the handler and converts it to `Data`.

```Swift
router.get("/users/:id") { request in
  return ["id": request.components["id"]!, "name": "Joan"]
}
```

Now everything is ready to test your mocked API; you can perform your request as you usually would do:

```Swift
let session = URLSession.shared
let url = URL(string: "http://www.test.com/users/1")!
session.dataTask(with: url) { (data, _, _) in
    // handle response
}.resume()
```

> Note: query parameters are not affecting the route match
> `http://www.test.com/users/1?foo=bar` would also be matched

In the previous example the handler was returning a simple `Dictionary`; while this works because `Dictionary` is already `Serializable`, you can also create your own entities that conform to `Serializable`:

```Swift
struct User: Serializable {
    let firstName: String
    let lastName: String
    let id: String
}

router.get("/users/:id") { request in
  return User(firstName: "Joan", lastName: "Romano", id: request.components["id"]!)
}
```

When a request is matched, the RouteHandler receives a `Request` object that represents your request including components, query parameters, httpBody and httpHeaders. The `Request` object can be useful when building dynamic responses.

#### Third-Party Libraries

Third-Party libraries that use the Foundation networking APIs are also supported but you might need to set a proper `URLSessionConfiguration`.  
For example, to setup `Alamofire`:

```swift
let configuration = URLSessionConfiguration.default
configuration.protocolClasses = [Server.self]
let sessionManager = SessionManager(configuration: configuration)
```

### Leverage the Store - Dynamic mocking

Kakapo gets even more powerful when using your Routers together with the Store. You can create, insert, remove, update or find objects.

This lets you mock the APIs behaviors as if you were using a real backend. This is the **dynamic** side of Kakapo.

To create entities that can be used with the store, your types need to conform to the `Storable` protocol.

```Swift
struct Article: Storable, Serializable {
    let id: String
    let text: String

    init(id: String, store: Store) {
        self.id = id
        self.text = randomString() // you might use some faker library like Fakery!
    }
}
```

An example usage could be to retrieve a specific `Article`:

```Swift
let store = Store()
store.create(Article.self, number: 20)

router.get("/articles/:id") { request in
  let articleId = request.components["id"]!
  return store.find(Article.self, id: articleId)
}
```

Of course you can perform any logic which fits your needs:

```Swift
router.post("/article/:id") { request in
    return store.insert { (id) -> Article in
        return Article(id: id, text: "text from the body")
    }
}

router.del("/article/:id") { request in
  let articleId = request.components["id"]!
  let article = store.find(Article.self, id: articleId)!
  try! store.delete(article)

  return ["status": "success"]
}
```

### CustomSerializable

In [Serializable](#serializable-protocol) we described how your classes can be serialized. The serialization, by default, will `Mirror` (using Swift's reflection) an entity by recursively serializing its properties.

Whenever a different behavior is needed, you can instead conform to `CustomSerializable` to provide your custom serialization.

For instance, `Array` uses `CustomSerializable` to return an `Array` containing its serialized elements. `Dictionary`, similarly, is serialized by creating a `Dictionary` with the same keys and serialized values.

For other examples of `CustomSerializable` and how to use it to create more complex serializations, take a look at the `JSONAPISerializer` implementation.

### JSONAPI

Since Kakapo was built with JSONAPI support in mind, `JSONAPISerializer` is able to serialize your entity into JSON conforming to [jsonapi.org](http://jsonapi.org).

Your entities, in order to be serialized conforming to JSONAPI, need to conform to `JSONAPIEntity` protocol.  

Let's see an example:

```Swift
struct Cat: JSONAPIEntity {
    let id: String
    let name: String
}

struct User: JSONAPIEntity {
    let id: String
    let name: String
    let cats: [Cat]
}
```

Note that `JSONAPIEntity` objects are already `Serializable` and you could just use them together with your Routers. However, to completely follow the JSONAPI structure in your responses, you should wrap them into a `JSONAPISerializer` struct:

```Swift
router.get("/users/:id") { request in
  let cats = [Cat(id: "33", name: "Joan"), Cat(id: "44", name: "Hez")]
  let user = User(id: "11", name: "Alex", cats: cats)
  return JSONAPISerializer(user)
}
```

### Expanding Null values with Property Policy

When serializing to JSON, you may want to represent a property value as `null`. For this, you can use the `PropertyPolicy` enum. It is similar to `Optional`, providing an additional `.null` case:

```Swift
public enum PropertyPolicy<Wrapped>: CustomSerializable {
    case None
    case Null
    case Some(Wrapped)
}
```

It's only purpose is to be serialized in 3 different ways, to cover all possible behaviors of an Optional property.
`PropertyPolicy` works exactly as `Optional` properties:
- `.none` -> property not included in the serialization
- `.some(wrapped)` -> serialize `wrapped`

The additional case ,`.null`, is serialized as `null` when converted to json.

```Swift
PropertyPolicy<Int>.none.serialized() // nil
PropertyPolicy<Int>.null.serialized() // NSNull
PropertyPolicy<Int>.some(1).serialized() // 1
```

### Key customization - Serialization Transformer

The keys of the JSON generated by the serialization are directly reflecting the property names of your entities. However, you might need different behaviors. For instance, many APIs use `snake_case` keys but almost everyone use `camelCase` properties in Swift.  
To transform the keys you can use `SerializationTransformer`. Objects conforming to this protocol are able to transform the keys of a wrapped object at serialization time.  

For a concrete implementation, check `SnakecaseTransformer`: a struct that implements `SerializationTransformer` to convert keys into snake case:

```Swift
let user = User(userName: "Alex")
let serialized = SnakecaseTransformer(user).serialized()
print(serialized) // [ "user_name" : "Alex" ]
```

### Customize responses with ResponseFieldsProvider

If your responses need to specify status code (which will be 200 by default) and/or header fields, you can take advantage of `ResponseFieldsProvider` to customize your responses.

Kakapo provides a default `ResponseFieldsProvider` implementation in the Response struct, which you can use to wrap your Serializable objects:

```Swift
router.get("/users/:id"){ request in
    return Response(statusCode: 400, body: user, headerFields: ["access_token" : "094850348502"])
}

let url = URL(string: "http://www.test.com/users/2")!
session.dataTaskWithURL() { (data, response, _) in
    let allHeaders = response.allHeaderFields
    let statusCode = response.statusCode
    print(allHeaders["access_token"]) // 094850348502
    print(statusCode) // 400
}.resume()
```

Otherwise your `Serializable` object can directly implement the protocol: take a look at `JSONAPIError` to see another example.

## Roadmap

Even though Kakapo is ready to use, it is not meant to be shipped to the App Store although you can also do it! In fact, you might see it in action in some Apple stores since it was used to mock some features of Runtastic's demo app; however, it's at its early stage and we would love to hear your thoughts. We encourage you to open an issue if you have any questions, feedbacks or you just want to propose new features.

- Full JSON API support [#67](https://github.com/devlucky/Kakapo/issues/67)
- Reverse and Recursive relationships [#16](https://github.com/devlucky/Kakapo/issues/16)
- Custom Serializers for common json specifications

## Examples

### Newsfeed [![BuddyBuild](https://dashboard.buddybuild.com/api/statusImage?appID=57e58ce073e94e0100c34a01&branch=master&build=latest)](https://dashboard.buddybuild.com/apps/57e58ce073e94e0100c34a01/build/latest)

Make sure you check the [demo app](https://github.com/devlucky/Kakapo/tree/master/Examples/NewsFeed) we created using Kakapo: a prototyped newsfeed app which lets the user create new posts and like/unlike them.  
To quickly try it use: `pod try Kakapo`

![](https://raw.githubusercontent.com/devlucky/Kakapo/master/Examples/NewsFeed/newsfeed.png)

## Authors

[@MP0w](https://github.com/MP0w) - [@zzarcon](https://github.com/zzarcon) - [@joanromano](https://github.com/joanromano)


Changelog.md
# Changelog

### HEAD
--------------

- ***Swift 4.0 Support***
- Fix handler was called even when the request was cancelled

### 2.1.0
-----------

- Added `PATCH` method support [Hernan Gonzalez](https://github.com/hernangonzalez)

### 2.0.0
-----------

#### Breaking

- ***Swift 3.0 Support***
- Renamed `HTTPBody` to `httpBody` and `HTTPHeader` to `httpHeader`
- `HTTPMethod` enum cases are now lowercase
- Updated APIs to follow Swift 3 new naming guidelines:

  #### Serializable
  - `serialize(_ keyTransformer: KeyTransformer? = nil) -> Any?` -> `serialized(transformingKeys keyTransformer: KeyTransformer? = nil) -> Any?`

  #### CustomSerializable
  - `customSerialize(_ keyTransformer: KeyTransformer?) -> Any?` -> `customSerialized(transformingKeys keyTransformer: KeyTransformer?) -> Any?`

  #### Store
  - `filter<T: Storable>(_: T.Type, includeElement: (T) -> Bool) -> [T]` -> `filter<T: Storable>(_: T.Type, isIncluded: (T) -> Bool) -> [T]`

  #### JSONAPISerializable
  - `data(includeRelationships: Bool, includeAttributes: Bool, keyTransformer: KeyTransformer?) -> Any?` -> `data(includingRelationships: Bool, includingAttributes: Bool, transformingKeys keyTransformer: KeyTransformer?) -> Any?`
  - `includedRelationships(includeChildren: Bool, keyTransformer: KeyTransformer?) -> [Any]?` -> `includedRelationships(includingChildren: Bool, transformingKeys keyTransformer: KeyTransformer?) -> [Any]?`

  #### JSONAPISerializer
  - `init(_ object: T, topLevelLinks: [String: JSONAPILink]? = nil, topLevelMeta: Serializable? = nil, includeChildren: Bool = false)` -> `init(_ object: T, topLevelLinks: [String: JSONAPILink]? = nil, topLevelMeta: Serializable? = nil, includingChildren: Bool = false)`


### 1.0.1
-----------

- Fix `Router` to handle same url with different HTTP methods

### 1.0.0
------------

- Swift 2.3 support
- Renamed `KakapoDB` to `Store`
- `init(id:db:)` required by `Storable` protocol has been changed to `init(id:store:)`
- Renamed `KakapoServer` to `Server`

### 0.2.0
------------

- Just another Swift 2.2 release before 1.0.0
- Implement `NSURLProtocol.stopLoading()` for delayed requests (#96) by @leviathan
- Prevent empty include array on `JSONAPISerializable arrays when no relationships
- Update excluded link key `topLinks` to `relationshipsLinks`

### 0.1.0
------------

- Initial release 🎉


Source/JSONAPILinks.swift
//
//  JSONAPILinks.swift
//  Kakapo
//
//  Created by Joan Romano on 19/06/16.
//  Copyright © 2016 devlucky. All rights reserved.
//

import Foundation

/**
 * An enum representing JSON API Link types.
 
 * A link MUST be represented as either:
 
    - A string containing the link’s URL.
    - An object which can contain the following members:
        - `href`: a string containing the link’s URL.
        - `meta`: a meta object containing non-standard meta-information about the link.
 
 * [See the JSON API documentation on links](http://jsonapi.org/format/#document-links)
 */
public enum JSONAPILink: CustomSerializable {

    /** A string containing the link’s URL. */
    case simple(value: String)
    /**
     An object which can contain more information than the link itself
     
     - parameter href: a string containing the link’s URL.
     - parameter meta: a meta object containing non-standard meta-information about the link.
     */
    case object(href: String, meta: Serializable)
    
    // MARK: CustomSerializable
    
    /**
     The `JSONAPILink` implementation of `CustomSerializable` returns directly the link for `.simple` or returns a dictionary containing the link (href key) and the serialized meta object (meta key) for `.object`
     */
    public func customSerialized(transformingKeys keyTransformer: KeyTransformer?) -> Any? {
        switch self {
        case let .object(href, meta):
            var serializedObject: [String: Any] = ["href": href]
            serializedObject["meta"] = meta.serialized(transformingKeys: keyTransformer)
            
            return serializedObject
        case let .simple(value):
            return value
        }
    }
}

/**
 * A protocol representing a JSON API entity with links. This will handle all different `links` combinations:
 
 - Inside `data` field
 - Inside `included` field
 - Inside `relationships` field
 
 * For the first two options, an entity must use the `links` property. In order to add links in the `relationships` field,
 an entity must use `relationshipsLinks` property and return them for every relationship. For example, given the representation:
 
 ```swift
 struct Cat: JSONAPIEntity, JSONAPILinkedEntity {
    let id: String
    let name: String
    let links: [String : JSONAPILink]?
 }
 
 struct Dog: JSONAPIEntity, JSONAPILinkedEntity {
    let id: String
    let name: String
    let links: [String : JSONAPILink]?
 }
 
 struct User: JSONAPIEntity, JSONAPILinkedEntity {
    let id: String
    let name: String
    let cats: [Cat]
    let links: [String : JSONAPILink]?
    let relationshipsLinks: [String : [String : JSONAPILink]]?
 }
 ```
 
 Apart from their own entity links, and in order to provide extra links for relationships, `User` must specify them for each relationship key:
 
 ```swift
 let cats = [Cat(id: "33", name: "Stancho", links: nil),
             Cat(id: "44", name: "Hez", links: ["test": JSONAPILink.Simple(value: "hello"), 
                                                "another": JSONAPILink.Object(href: "http://example.com/articles/1/comments", meta: Meta())])]
 
 let dog = Dog(id: "22", name: "Joan", links: nil)
 
 let user = User(id: "11", name: "Alex", dog: dog, cats: cats, 
                 links: ["one": JSONAPILink.Simple(value: "hello"),
                         "two": JSONAPILink.Object(href: "hello", meta: Meta())],
                 relationshipsLinks: ["cats": ["prev": JSONAPILink.Simple(value: "hello"),
                                               "next": JSONAPILink.Simple(value: "world"),
                                               "first": JSONAPILink.Simple(value: "yeah"),
                                               "last": JSONAPILink.Simple(value: "text")],
                                      "dog": ["testDog": JSONAPILink.Simple(value: "hello"),
                                              "anotherDog": JSONAPILink.Object(href: "http://example.com/articles/1/comments", meta: Meta())]
                                     ]
            )
 ```
 
 * In order to represent the top level links of the root object, check out `JSONAPISerializer` initialization methods.
 
 * [See the JSON API documentation on links](http://jsonapi.org/format/#document-links)
 */
public protocol JSONAPILinkedEntity {
    /// The related links, must use link-names as keys and links as values.
    var links: [String: JSONAPILink]? { get }
    /// The relationships links, an object containing relationships can specify top level links for every relationship type. The object must provide a Dictionary where keys are the relationships types and values are dictionaries with link-names as keys and link as values.
    var relationshipsLinks: [String: [String: JSONAPILink]]? { get }
}

extension JSONAPILinkedEntity {
    /// Default Implementation returns nil
    public var links: [String: JSONAPILink]? { return nil }
    /// Default Implementation returns nil
    public var relationshipsLinks: [String: [String: JSONAPILink]]? { return nil }
}
